browser/ai_chat/content_agent_tool_provider.cc

```diff
@@ -0,0 +1,183 @@
+}
+
+ContentAgentToolProvider::~ContentAgentToolProvider() = default;
+
+std::vector<base::WeakPtr<Tool>> ContentAgentToolProvider::GetTools() {
+  // Note: We don't have the ability to filter tools based on conversation
+  // capability here. But for now we don't need to as we only create the content
+  // agent tools if we don't have the actor_service_ (which is only provided for
+  // agent profiles).
+  std::vector<base::WeakPtr<Tool>> tool_ptrs;
+  tool_ptrs.reserve(tools_.size());
+  for (const auto& tool : tools_) {
+    tool_ptrs.push_back(tool->GetWeakPtr());
+  }
+  return tool_ptrs;
+}
+
+void ContentAgentToolProvider::StopAllTasks() {
+  if (actor_service_ && !task_id_.is_null()) {
+    // `success` sets whether the task ends as state kFinished or kCancelled
+    actor_service_->StopTask(task_id_, true /* success */);
+  }
+}
+
+actor::TaskId ContentAgentToolProvider::GetTaskId() {
+  return task_id_;
+}
+
+void ContentAgentToolProvider::GetOrCreateTabHandleForTask(
+    base::OnceCallback<void(tabs::TabHandle)> callback) {
+  if (!task_tab_handle_.Get()) {
+    // Create a new tab because we are only allowed to act on
+    // certain URLs, e.g. NTP. Safer to start on a blank page
+    // whilst this feature is focused on AI-initiated tasks instead
+    // of acting on existing tabs.
+    NavigateParams params(profile_, GURL(url::kAboutBlankURL),
+                          ui::PAGE_TRANSITION_FROM_API);
+    params.disposition = WindowOpenDisposition::NEW_BACKGROUND_TAB;
+    Navigate(&params);
+    content::WebContents* new_contents = params.navigated_or_inserted_contents;
+
+    task_tab_handle_ =
+        tabs::TabInterface::GetFromContents(new_contents)->GetHandle();
+  }
+  actor_service_->GetTask(task_id_)->AddTab(
+      task_tab_handle_, base::BindOnce(&ContentAgentToolProvider::TabAdded,
+                                       weak_ptr_factory_.GetWeakPtr(),
+                                       task_tab_handle_, std::move(callback)));
+}
+
+void ContentAgentToolProvider::TabAdded(
+    tabs::TabHandle tab_handle,
+    base::OnceCallback<void(tabs::TabHandle)> callback,
+    actor::mojom::ActionResultPtr result) {
+  std::move(callback).Run(tab_handle);
+}
+
+void ContentAgentToolProvider::ExecuteActions(
+    optimization_guide::proto::Actions actions,
+    Tool::UseToolCallback callback) {
+  actor::BuildToolRequestResult requests = actor::BuildToolRequest(actions);
+
+  if (!requests.has_value()) {
+    DLOG(ERROR) << "Action Failed to convert BrowserAction to ToolRequests.";
+    std::move(callback).Run(
+        CreateContentBlocksForText("Action failed - incorrect parameters"));
+    return;
+  }
+
+  actor_service_->PerformActions(
+      actor::TaskId(actions.task_id()), std::move(requests.value()),
+      base::BindOnce(&ContentAgentToolProvider::OnActionsFinished,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void ContentAgentToolProvider::CreateTools() {
+  tools_.clear();
```

Comment:
why do we need to clear the tools here? This method is only called from the constructor (and its private)

Reply:
A fair observation. I did it because we're not checking the existance of any of the tools or being intelligent about when to add each one. It's possible we may want to re-create the tools during `ToolProvider::OnNewGenerationLoop()` if the tools hang on to state. I might have had that happen previously and removed it. Is there a disadvantage of leaving the clear here?

Reply:
I guess not, it just read as slightly jarring given its only called from the ctor